manifests/pkcs11_eventmgr.pp

```diff
@@ -32,46 +32,63 @@
 #
 #   Default: {}
 #
+# @param event_opts_base
+#   Hash
+#
+#   Default: $pam_pkcs11::params::pkcs11_event_opts
+#
 # @param autostart_method
 #   String
 #
 #   Default: $pam_pkcs11::params::pkcs11_eventmgr_autostart_method,
 #
 class pam_pkcs11::pkcs11_eventmgr (
-  Boolean    $debug            = false,
-  Boolean    $daemonize        = true,
-  Integer[0] $polling_time     = 1,
-  Integer[0] $expire_time      = 0,
-  String     $pkcs11_module    = 'default',
-  Hash       $event_opts       = {},
-  String     $autostart_method = $pam_pkcs11::params::pkcs11_eventmgr_autostart_method,
+  Boolean                 $debug            = false,
+  Boolean                 $daemonize        = true,
+  Integer[0]              $polling_time     = 1,
+  Integer[0]              $expire_time      = 0,
+  Variant[Enum['default'],
+  Stdlib::AbsolutePath] $pkcs11_module      = 'default',
+  Struct[
+    card_insert => Optional[Struct[
+        on_error => Enum['ignore', 'return', 'quit'],
+        action => Array[String],
+    ]],
+    card_remove => Optional[Struct[
+        on_error => Enum['ignore', 'return', 'quit'],
+        action => Array[String],
+    ]],
+    expire_time => Optional[Struct[
+        on_error => Enum['ignore', 'return', 'quit'],
+        action => Array[String],
+  ]]]                     $event_opts       = {},
+  Struct[
+    card_insert => Struct[
+      on_error => Enum['ignore', 'return', 'quit'],
+      action => Array[String],
+    ],
+    card_remove => Struct[
+      on_error => Enum['ignore', 'return', 'quit'],
+      action => Array[String],
+    ],
+    expire_time => Struct[
+      on_error => Enum['ignore', 'return', 'quit'],
+      action => Array[String],
+  ]]                      $event_opts_base  = $pam_pkcs11::params::pkcs11_event_opts,
```

Comment:
There is quite some duplication here. How about we create [type aliases](https://puppet.com/docs/puppet/7/lang_type_aliases.html) for both the outer and inner structs?

Reply:
Yes I thought about this as well.. But one of them contain 'Optional' structs while the other everything is required.

Reply:
I will look how I can still make this a bit easier for the eyes and remove the redundancy as much as possible...

Reply:
Ah, right. I guess we could definitely have the type alias for the inner struct but the outer would be more problematic. I didn't find a way to take a struct and create a new type where all attributes are optional.

Some further discussions [below](https://github.com/siemens/puppet-pam_pkcs11/pull/1#discussion_r929856012).